daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -185,96 +224,84 @@ class VariableMap private(vTable: Map[GlobalQName, List[List[Variable]]])
     val varQName = vrd.globalQName
+    val stack = vTable.get(varQName)
+    if (stack.isDefined) {
+      val variable = stack.get.top
+      variable.state match {
+        case VariableSet => {
+          referringContext.SDE("Cannot set variable %s twice. State was: %s. Existing value: %s",
+          variable.rd.globalQName, VariableSet, variable.value)
+        }
 
-    vTable.get(varQName) match {
-
-      case None => referringContext.schemaDefinitionError("unknown variable %s", varQName)
-
-      // There should always be a list with at least one tier in it (the global tier).
-      case x @ Some(firstTier :: enclosingScopes) => {
-        firstTier match {
-
-          case Variable(VariableDefined, v, ctxt, defaultExpr) :: rest if (v.isDefined) => {
-            val newVar = Variable(VariableSet, VariableUtils.convert(newValue.getAnyRef.toString, ctxt), ctxt, defaultExpr)
-            mkVMapForNewVariable(newVar, firstTier, enclosingScopes)
-          }
-
-          case Variable(VariableUndefined, DataValue.NoValue, ctxt, defaultExpr) :: rest => {
-            val newVar = Variable(VariableSet, VariableUtils.convert(newValue.getAnyRef.toString, ctxt), ctxt, defaultExpr)
-            mkVMapForNewVariable(newVar, firstTier, enclosingScopes)
-          }
-
-          case Variable(VariableSet, v, ctxt, defaultExpr) :: rest if (v.isDefined) => {
-            referringContext.SDE("Cannot set variable %s twice. State was: %s. Existing value: %s", ctxt.globalQName, VariableSet, v)
-          }
-
-          case Variable(VariableRead, v, ctxt, defaultExpr) :: rest if (v.isDefined) => {
-            // referringContext.SDE
-            pstate.SDW("Cannot set variable %s after reading the default value. State was: %s. Existing value: %s", ctxt.globalQName, VariableSet, v)
-            val newVar = Variable(VariableSet, VariableUtils.convert(newValue.getAnyRef.toString, ctxt), ctxt, defaultExpr)
-            mkVMapForNewVariable(newVar, firstTier, enclosingScopes)
-          }
+        case VariableRead => {
+          referringContext.SDE("Cannot set variable %s after reading the default value. State was: %s. Existing value: %s",
+          variable.rd.globalQName, VariableSet, variable.value)
+        }
 
-          case _ => Assert.invariantFailed("variable map internal list structure not as expected: " + x)
+        case _ => {
+          variable.setValue(VariableUtils.convert(newValue.getAnyRef.toString, variable.rd))
+          variable.setState(VariableSet)
         }
       }
-      case x => Assert.invariantFailed("variables data structure not as expected. Should not be " + x)
     }
   }
 
+  /**
+   * Creates a new instance of a variable
+   */
+  def newVariableInstance(vrd: VariableRuntimeData) = {
+    val varQName = vrd.globalQName
+    val stack = vTable.get(varQName)
+    if (stack.isDefined) {
+      stack.get.push(vrd.newVariableInstance)
+    }
```

Comment:
In what case is stack not defined? Should this be an Assertion that stack.isDefined instead